### PR TITLE
API-4394: Changed Create Topic endpoint from POST to PUT and code so …

### PR DESCRIPTION
**Pull request by mattclark-zerogravit**
…if topic already exists we just return the topicId and not an error

### CODE REVIEW

**Comment by billymitchell14** [comment]
👍

File: README.md [hunk 1]
```diff
@@ -79,7 +79,12 @@ This endpoint is restricted, only a whitelist of services can access it.
 | `clientId` | The Developer Hub Client ID that can access this topic
 
 ### Response
-HTTP Status: `201`
+HTTP Status: `201` if topic is created
+```
+1c5b9365-18a6-55a5-99c9-83a091ac7f26
```
Comment (peteslater-ee) [inline README.md:84]: Should this be a JSON document example?

File: app/uk/gov/hmrc/pushpullnotificationsapi/controllers/TopicsController.scala [hunk 2]
```diff
@@ -48,22 +49,23 @@ class TopicsController @Inject()(appConfig: AppConfig,
       .async(playBodyParsers.json) { implicit request =>
         withJsonBody[CreateTopicRequest] {
           topic: CreateTopicRequest =>
-            val topicId = UUID.randomUUID().toString
-            topicsService.createTopic(topicId, topic.clientId, topic.topicName).map { _ =>
-              Logger.info(s"Topic Created: $topicId for clientId: ${topic.clientId}")
-              Created(topicId)
-            } recover recovery
-        }
+            val topicId = TopicId(UUID.randomUUID())
```
Comment (peteslater-ee) [inline app/uk/gov/hmrc/pushpullnotificationsapi/controllers/TopicsController.scala:52]: Could we get collisions with existing UUIDs here?

File: app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala [hunk 3]
```diff
@@ -39,10 +44,16 @@ object ResponseFormatters{
   implicit val JodaDateReads: Reads[org.joda.time.DateTime] = JodaReads.jodaDateReads(dateFormat)
   implicit val JodaDateWrites: Writes[org.joda.time.DateTime] = JodaWrites.jodaDateWrites(dateFormat)
   implicit val JodaDateTimeFormat: Format[DateTime] = Format(JodaDateReads, JodaDateWrites)
-  implicit val notificationWrites = Json.format[Notification]
+  implicit val notificationIdFormatter: OFormat[NotificationId] = Json.format[NotificationId]
+  implicit val topicIdFormatter: OFormat[TopicId] = Json.format[TopicId]
+  implicit val notificationWrites: OFormat[Notification] = Json.format[Notification]
```
Comment (peteslater-ee) [inline app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala:49]: Probably a formatter now and not a writes

File: app/uk/gov/hmrc/pushpullnotificationsapi/models/models.scala [hunk 4]
```diff
@@ -21,13 +21,30 @@ import java.util.UUID
...
+object SubscriberId{
+  def fromString(value: String)={
+    SubscriberId(UUID.fromString(value))
+  }
+}
+
+case class TopicCreator(clientId: ClientId)
 
 object TopicCreator {
+  implicit val clientIdFormatter: OFormat[ClientId] = Json.format[ClientId]
```
Comment (peteslater-ee) [inline app/uk/gov/hmrc/pushpullnotificationsapi/models/models.scala:47]: Is this the same one as in JsonFormatters?

File: app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala (hunk 3 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala:49]: OK will rename

File: app/uk/gov/hmrc/pushpullnotificationsapi/controllers/TopicsController.scala (hunk 2 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/controllers/TopicsController.scala:52]: Yeah but as its UUID v4 the chance is very very low.

File: README.md (hunk 1 above)
Comment (mattclark-zerogravit) [reply README.md:84]: The read me is being covered off in a separate document ticket API-4321

File: app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala (hunk 3 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/models/jsonformatters.scala]: done

File: app/uk/gov/hmrc/pushpullnotificationsapi/models/models.scala (hunk 4 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/models/models.scala]: done

**Pull request merged**